ipa-core/src/protocol/ipa_prf/oprf_padding/mod.rs

```diff
@@ -1,5 +1,866 @@
+pub async fn apply_dp_padding_pass<C, T, const B: usize>(
+    ctx: C,
+    mut input: Vec<T>,
+    h_i: Role,
+    h_i_plus_one: Role,
+    h_out: Role,
+    padding_params: &PaddingParameters,
+) -> Result<Vec<T>, Error>
+where
+    C: Context,
+    T: Paddable,
+{
+    // assert roles are all unique
+    assert!(h_i != h_i_plus_one);
+    assert!(h_i != h_out);
+    assert!(h_out != h_i_plus_one);
+
+    let mut total_number_of_fake_rows = 0;
+    let mut padding_input_rows: Vec<T> = Vec::new();
+
+    // Step 1: Helpers `h_i` and `h_i_plus_one` will get the same rng from PRSS
+    // and use it to sample the same random noise for padding from OPRFPaddingDp.
+    // They will generate secret shares of these fake rows.
+    if ctx.role() != h_out {
+        total_number_of_fake_rows = two_parties_add_dummies::<C, T, B>(
+            &ctx,
+            &mut padding_input_rows,
+            h_i,
+            h_i_plus_one,
+            padding_params,
+        )?;
+    }
+
+    // Step 2: h_i and h_i_plus_one will send the send total_number_of_fake_rows to h_out. h_out will
+    // check that both h_i and h_i_plus_one have sent the same value to prevent any malicious behavior
+    // See oprf_padding/README.md for explanation of why revealing the total number of fake rows is okay.
+    let send_ctx = ctx
+        .narrow(&SendTotalRows::SendNumFakeRecords)
+        .set_total_records(TotalRecords::ONE);
+    match ctx.role() {
+        role if role == h_i || role == h_i_plus_one => {
+            let direction = if role == h_i {
+                Direction::Left
+            } else {
+                Direction::Right
+            };
+            let send_channel = send_ctx.send_channel::<BA32>(send_ctx.role().peer(direction));
+            let _ = send_channel
+                .send(
+                    RecordId::FIRST,
+                    BA32::truncate_from(u128::from(total_number_of_fake_rows)),
+                )
+                .await;
+        }
+        _h_out => {
+            let recv_channel_right =
+                send_ctx.recv_channel::<BA32>(send_ctx.role().peer(Direction::Right));
+            let recv_channel_left =
+                send_ctx.recv_channel::<BA32>(send_ctx.role().peer(Direction::Left));
+            let (from_right, from_left) = try_join!(
+                async {
+                    match recv_channel_right.receive(RecordId::FIRST).await {
+                        Ok(v) => Ok::<u32, Error>(u32::try_from(v.as_u128()).unwrap()),
+                        Err(e) => Err(e.into()),
+                    }
+                },
+                async {
+                    match recv_channel_left.receive(RecordId::FIRST).await {
+                        Ok(v) => Ok::<u32, Error>(u32::try_from(v.as_u128()).unwrap()),
+                        Err(e) => Err(e.into()),
+                    }
+                }
+            )?;
+
+            assert_eq!(from_right, from_left);
```

Comment:
An assert will crash the server, so the rule of thumb for asserts is that they shouldn't be reachable based on external input. It would be better to propagate this as an error, maybe `Error::InconsistentShares` (which should probably have a member to provide some context about what shares where inconsistent, maybe the name of the function throwing it.

Reply:
okay, created an `Error::InconsistentPadding` to return here.